File: projects/m5/001-budget-app/budget.py
class Category:
    def __init__(self, category_name):
        self.category_name = category_name
        self.ledger = []
    
    def deposit(self, amount, description=""):
        if isinstance(amount, (int, float)):
            self.ledger.append({"amount": amount, "description": description})
        else:
            raise TypeError("The amount parameter can be an int or a float.")

    def withdraw(self, amount, description=""):
        if isinstance(amount, (int, float)):
            if self.check_funds(amount):
                self.ledger.append({"amount": -amount, "description": description})
                return True
            return False
        raise TypeError("The amount parameter can be an int or a float.")

    def get_balance(self):
        current_balance = 0
        for move in self.ledger:
            current_balance += float(move["amount"])
        return current_balance
    
    def transfer(self, amount, another_category):
        if isinstance(amount, (int, float)):
            if self.check_funds(amount):
                self.withdraw(amount, "Transfer to {}".format(another_category.category_name))
                another_category.deposit(amount, "Transfer from {}".format(self.category_name))
                return True
            return False
        raise TypeError("The amount parameter can be an int or a float.")

    def check_funds(self, amount):
        if isinstance(amount, (int, float)):
            current_balance = self.get_balance()
            if amount > current_balance:
                return False
            return True
        raise TypeError("The amount parameter can be an int or a float.")

    def __str__(self):
        printed_object = "{:*^30}\n".format(self.category_name)
        for item in self.ledger:
            printed_object += "{:23.23}{:7.2f}\n".format(item["description"], item["amount"])
        printed_object += "Total: {:.2f}".format(self.get_balance())
        return printed_object


def create_spend_chart(categories: list) -> str:
    """Takes a list of categories, returns a string which is a bar chart."""
    chart = ""
    total_expenses = 0
    expenses_by_category = dict()
    percentages_by_category = dict()
 
    for single_category in categories:
        expenses_by_category[single_category.category_name] = 0
        for movement in single_category.ledger:
            if movement["amount"] < 0:
                total_expenses -= movement["amount"]
                expenses_by_category[single_category.category_name] -= movement["amount"]

    for element in expenses_by_category:
        percentage = expenses_by_category[element] / total_expenses * 100
        percentages_by_category[element] = percentage

    title = "Percentage spent by category\n"
    left_side = {100: "", 90: "", 80: "", 70: "", 60: "", 50: "", 40: "", 30: "", 20: "", 10: "", 0: ""}
    for single_category in categories:
        for number in left_side:
            if percentages_by_category[single_category.category_name] > number:
                left_side[number] += " o "
            else:
                left_side[number] += "   "
    
    dashes_number = 3 * len(categories) + 1
    dashed_line = "    " + ("-" * dashes_number) + "\n"

    category_name_rows = dict()
    index_category = 1
    for single_category in categories:
        row = 1
        for letter in single_category.category_name:
            if row not in category_name_rows:
                category_name_rows[row] =  (3 * index_category * " ") + " {} ".format(letter)
            else:
                category_name_rows[row] += " {} ".format(letter)
            row += 1
        if len(single_category.category_name) < max(category_name_rows.keys()):
            while row <= max(category_name_rows.keys()):
                category_name_rows[row] += "   "
                row += 1
        index_category += 1
    
    chart += title
    for value in reversed(sorted(left_side.keys())):
        chart += "{:3}|{} \n".format(value, left_side[value])
    chart += dashed_line

    for row_number in sorted(category_name_rows.keys()):
        chart += " {} ".format(category_name_rows[row_number])
        if row_number < max(category_name_rows.keys()):
            chart += "\n"
            
    return chart

File: projects/m5/001-budget-app/js/src/budget.js
// @ts-check
/**
 * @file Create a Category class.
 * @module budget
 * @author Salvatore Quagliariello
 */


/** 
 * A Class to create new categories for a Budget App.
 */
class Category {
    /**
     * Category details.
     * @param {string} name
     */
    constructor(name) {
        /**
         * @property {string} name The name of the Category Object.
         */
        this.name = name;
        /**
         * @property {Array<Object>} ledger An Array containing every transaction made.
         */
        this.ledger = [];
    }

    /**
     * Create an object containing details about a deposit transaction and push it to the `ledger` property.
     * @param {number} amount The amount deposited.
     * @param {string} description The description for the transaction.
     */
    deposit(amount, description = ``) {
        this.ledger.push({description, amount});
    }
    
    /**
     * Create an Object containing details about a withdraw transaction and push it to the `ledger` property.
     * @param {number} amount The amount withdrawed.
     * @param {string} description The description for the transaction.
     * @returns {boolean} Returns `true` if the withdraw has been successful, otherwise return `false`.
     */
    withdraw(amount, description = ``) {
        if (!this.check_funds(amount)) return false;

        this.ledger.push({description, amount: -amount})
        return true;
    }

    /**
     * Return the total balance, given by the sum of every transaction made.
     * @returns {number}
     */
    get_balance() {
        const balance = this.ledger.reduce((n, {amount}) => n + amount, 0);
        return balance;
    }

    /**
     * Return the total sum of every withdraw succesfully made.
     * @returns {number}
     */
    get_withdraws() {
        const withdraws = this.ledger.filter(obj => obj.amount < 0);
        const sumOfWithdraws = withdraws.reduce((n, {amount}) => n + +amount, 0);
        return sumOfWithdraws;
    }

    /** 
     * Uses `this.check_funds` to check if the amount parameter is greater than the balance of 
     * the Object this method is applied. If so it uses the `withdraw` method on the applied Object
     * and the `deposit` method on the parameter Object to create a transfer transaction.
     * @param {number} amount The amount to transfer.
     * @param {Object} category The Category Object to transfer the given amount.
     * @returns {boolean} Return `true` is the transfer transaction has been succesfully completed.
     */
    transfer(amount, category) {
        if (!this.check_funds(amount)) return false;

        this.withdraw(amount, `Transfer to ${category.name}`);
        category.deposit(amount, `Transfer from ${this.name}`);
        return true;
    }

    /**
     * Check if the balance of the Object is greater than the given amount.
     * @param {number} amount 
     * @returns {boolean} Returns `true` if `balance` is greater than `amount`, otherwise return `false`.
     */
    check_funds(amount) {
        if (amount <= this.get_balance()){
            return true;
        } else {
            return false;
        }
    }

    /**
     * Create a graphic representation of the successfull transaction made on the Category. 
     * @returns {string}
     */
    print() {
        const asteriskNumber = 30 - this.name.length;
        const asterisks = Array(asteriskNumber / 2).fill(`*`).join(``);
        const title = `${asterisks}${this.name}${asterisks}`;
        let list = ``;
        
        for ( let obj of this.ledger) {
            const amountString = obj.amount.toFixed(2).toString().substring(0, 7);
            const descriptionString = obj.description.substring(0, 23);        
            const objectLength = descriptionString.length + amountString.length;
            const spaces = Array(30 - objectLength).fill(` `).join(``);
            const string = `${descriptionString}${spaces}${amountString}\n`;
            list += string;
        };
        
        const finalBalanceString = `Total: ${this.get_balance().toFixed(2)}`;
        const outputString = `${title}\n${list}${finalBalanceString}`;
        
        return outputString;
    }
}

module.exports = Category;